rocAL/rocAL_pybind/amd/rocal/plugin/pytorch.py

```diff
@@ -319,45 +318,81 @@ def __init__(self, pipeline, tensor_layout = types.NCHW, reverse_channels = Fals
         self.device_id = device_id
         self.output = None
         self.len = b.getRemainingImages(self.loader._handle)
-        # self.last_batch_policy = self.loader._last_batch_policy #commented for now
+
+        self.last_batch_policy = self.loader._last_batch_policy
         self.shard_size = size
         self.auto_reset = auto_reset
         self.batch_count = 0
-        self.audio_length = None
         self.samples = None
         self.channels = None
         self.max_shape = None
         self.batch_size = self.loader._batch_size
+        self.output_list = self.torch_dtype = None
+        self.labels_size = self.batch_size
+        self.output_memory_type = self.loader._output_memory_type
+
 
     def next(self):
         return self.__next__()
 
     def __next__(self):
-        if self.loader.rocalRun() != 0:
+        if self.loader.rocalRun() != 0 and self.shard_size < 0:
+            if self.auto_reset:
+                self.reset()
+            raise StopIteration
+
+        elif self.shard_size > 0 and self.batch_count >= self.shard_size :
+            if self.auto_reset:
+                self.reset()
             raise StopIteration
         else:
             self.output_tensor_list = self.loader.getOutputTensors()
 
+        self.last_batch_size = self.batch_size - b.getLastBatchPaddedSize(self.loader._handle)
         self.batch_count = self.batch_count + self.batch_size
         self.num_of_dims = self.output_tensor_list[0].num_of_dims()
-        if self.num_of_dims == 3:
-            self.batch_size = self.output_tensor_list[0].batch_size() if self.batch_size is None else self.batch_size
-            self.max_shape = self.output_tensor_list[0].max_shape() if self.max_shape is None else self.max_shape
-            self.channels = self.max_shape[0]
-            self.samples = self.max_shape[1]
-            self.audio_length = self.channels * self.samples if self.audio_length is None else self.audio_length
-            roi = self.output_tensor_list[0].get_rois().reshape(self.batch_size, 4)
-            max_x1 = np.max(roi[..., 0:1])
-            max_y1 = np.max(roi[..., 1:2])
-            self.output = torch.empty((self.batch_size, max_y1, max_x1), dtype=torch.float32)
-
-            self.labels = self.loader.getImageLabels()
-            self.labels_tensor = torch.from_numpy(self.labels).type(torch.LongTensor)
-
-            self.output_tensor_list[0].copy_data(ctypes.c_void_p(self.output.data_ptr()), max_y1, max_x1)
-            return self.output, self.labels_tensor, torch.tensor(roi[..., 0:2])
+
+        if self.output_list is None:
+            self.output_list = []
+            for i in range(len(self.output_tensor_list)):
```

Comment:
There is a overlap of code between the if and else of self.output_list
please try to modularize